hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/ContainerStateMachine.java

```diff
@@ -186,13 +186,39 @@ long getStartTime() {
     }
   }
 
+  static class WriteFutureContext {
+    private final CompletableFuture<ContainerCommandResponseProto> writeChunkFuture;
+    private final CompletableFuture<Message> raftFuture;
+    private final long startTime;
+
+    WriteFutureContext(CompletableFuture<ContainerCommandResponseProto> writeChunkFuture,
+                       CompletableFuture<Message> raftFuture, long startTime) {
+      this.writeChunkFuture = writeChunkFuture;
+      this.raftFuture = raftFuture;
+      this.startTime = startTime;
+    }
+
+    public CompletableFuture<ContainerCommandResponseProto> getWriteChunkFuture() {
+      return writeChunkFuture;
+    }
+
+    public CompletableFuture<Message> getRaftFuture() {
+      return raftFuture;
+    }
+
+    long getStartTime() {
+      return startTime;
+    }
+  }
+
   private final SimpleStateMachineStorage storage =
       new SimpleStateMachineStorage();
   private final ContainerDispatcher dispatcher;
   private final ContainerController containerController;
   private final XceiverServerRatis ratisServer;
-  private final ConcurrentHashMap<Long,
-      CompletableFuture<ContainerCommandResponseProto>> writeChunkFutureMap;
+  private final ConcurrentHashMap<Long, WriteFutureContext> writeChunkFutureMap;
```

Comment:
We can make this concurrentSkipListMap instead and remove the writeFutureMinIndex. Keys would be in sorted order.

Reply:
We can just keep this variable as a SortedMap(Base interface), we need not keep it implementation specific